### PR TITLE
Task/password reset

### PR DESCRIPTION
**Pull request by cybernetlab**

### CODE REVIEW

File: lib/gpanel_web/controllers/password_reset_controller.ex [hunk 1]
```diff
@@ -1,12 +1,51 @@
...
-  alias Gpanel.Accounts
+  alias Gpanel.{Accounts, Email, Mailer}
+  alias GpanelWeb.LoadCurrentUser
+  alias Phoenix.Token
 
+  @token_secret Application.compile_env!(:gpanel, [GpanelWeb.Endpoint, :secret_key_base])
+  @token_salt "password-reset"
+  @token_ttl 86_400 # one week
+  @reset_token_salt "password-resetting"
+  @reset_token_ttl 900 # 15 minutes
```
Comment (Kintull) [inline lib/gpanel_web/controllers/password_reset_controller.ex:12]: Could you please explain how two kinds of ttl work together?

File: lib/gpanel_web/controllers/password_reset_controller.ex [hunk 2]
```diff
@@ -15,13 +54,52 @@ defmodule GpanelWeb.PasswordResetController do
         |> redirect(to: Routes.password_reset_path(conn, :show))
 
       user ->
-        #create token
-        #send email
+        token = Token.sign(@token_secret, @token_salt, user.id)
```
Comment (Kintull) [inline lib/gpanel_web/controllers/password_reset_controller.ex:57]: Link that is inserted in password reset mail can be used several times,do you think it is safe enough to leave it like this having only ttl check configured?

**Review by Kintull — CHANGES_REQUESTED** [review]

File: lib/gpanel_web/controllers/password_reset_controller.ex (hunk 1 above)
Comment (cybernetlab) [reply lib/gpanel_web/controllers/password_reset_controller.ex:12]: there are two kind of tokens:
1. Token that we receive as a parameter from password reset link when user clicks on emailed link (call it reset token).
2. After that we need to pass user-id to password-reset controller. It's not safe to use session or direct parameters for this. So we create another token (call it resetting token) and verifying it before reset password. This is temporary token to pass user-id between two controller actions.

File: lib/gpanel_web/controllers/password_reset_controller.ex (hunk 2 above)
Comment (cybernetlab) [reply lib/gpanel_web/controllers/password_reset_controller.ex:57]: password reset via email is not enough secured by design. I'm sure that we need to implement something like 2fa for this in future. Anyway I don't know other ways to authenticate user in this scenario.

File: lib/gpanel_web/templates/password_reset/reset.html.eex [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<div class="container">
+  <div class="row justify-content-md-center mt-3 pt-5">
+    <div class="col col-lg-4"></div>
+    <div class="col">
+      <%= form_for @changeset, Routes.password_reset_path(@conn, :update), [method: "put"], fn f -> %>
+        <input type="hidden" name="token" value="<%= @token %>" />
+
+        <div class="form-group row">
+          <div class="col-sm-12">
+            <input type="password" name="prev_password" class="register-form-control" placeholder="Previous password" required />
```
[the remark this answers is not in the document]
Comment (cybernetlab) [reply lib/gpanel_web/templates/password_reset/reset.html.eex:10]: haha, I've got carried away a bit )) Sure - where shouldn't be a prev password field

**Pull request merged**